sql/analyzer/apply_join.go

```diff
@@ -245,3 +246,153 @@ func normalizeSelectSingleRel(ctx *sql.Context, a *Analyzer, n sql.Node, scope *
 		return c.Node, transform.SameTree, nil
 	})
 }
+
+// hoistOutOfScopeFilters pulls filters upwards into the parent scope
+// to decorrelate subqueries for further optimizations.
+//
+// select * from xy where exists (select * from uv where x = 1)
+// =>
+// select * from xy where x = 1 and exists (select * from uv)
+func hoistOutOfScopeFilters(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
+	switch n.(type) {
+	case *plan.TriggerBeginEndBlock:
+		return n, transform.SameTree, nil
+	}
+	ret, same, _, err := recurseSubqueryForOuterFilters(n, a, scope)
+	return ret, same, err
+}
+
+// recurseSubqueryForOuterFilters recursively hoists filters that belong
+// to an outer scope (maybe higher than the parent). We do a DFS for hoisting
+// subquery filters. We do a BFS to extract hoistable filters from subquery
+// expressions before checking the normalized subquery and its hoisted
+// filters for further hoisting.
+func recurseSubqueryForOuterFilters(n sql.Node, a *Analyzer, scope *Scope) (sql.Node, transform.TreeIdentity, []sql.Expression, error) {
+	var hoistFilters []sql.Expression
+	lowestAllowedIdx := len(scope.Schema())
+	var inScope TableAliases
+	ret, same, err := transform.Node(n, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		sq, _ := n.(*plan.SubqueryAlias)
+		if sq != nil {
+			subScope := scope.newScopeFromSubqueryAlias(sq)
+			newQ, same, hoisted, err := recurseSubqueryForOuterFilters(sq.Child, a, subScope)
+			if err != nil {
+				return n, transform.SameTree, err
+			}
+			if same {
+				return n, transform.SameTree, nil
+			}
+			if len(hoisted) > 0 {
+				hoistFilters = append(hoistFilters, hoisted...)
+			}
+			ret := *sq
+			ret.Child = newQ
+			return &ret, transform.NewTree, nil
+		}
+		f, _ := n.(*plan.Filter)
```

Comment:
Nit: introduce and use `ok`variable, instead of checking `f == nil`.

Reply:
we can put this up for discussion in swes if you want. I've thought about this for awhile, I see a lot of usage of nil checks in other code bases, but have mostly used the ok pattern. I think I prefer only one variable allocation, faster and less surface area for errors.